py-rse/package-py.Rmd

```diff
@@ -1421,3 +975,166 @@ if (knitr::is_latex_output()) {
 
 ```{r, child="keypoints/rse-package-py.md"}
 ```
+
+## Exercises
+
+### Exercises: Examining the virtual environment
+
+1. If you haven't already, create a `test` virtual environment and activate it.  Your prompt should now start with `(test)`.
+2. Run `pip list`. What packages are installed?
+3. Run the `showpath.py` script from the earlier section, and note that the paths are different from your
+base environment where you looked for packages.  They probably look like `/Users/pterry/envs/test/lib/python-3.7/site-packages`.
+4. Pick a package you know you have installed in your base environment—pick `pytest` if you can't think of one—and
+try to import it in a Python interpreter.  Can it be found?
+5. Deactivate the environment and try again.
+
+### Exercise: Hello, Zipf!
+
+Put a statement
+
+```python
+print("Hello, Zipf!")
+```
+at the end of `zipf.py`.
+
+1. What happens when you start Python interactively and import the
+module with the command `import zipf`?
+2. Python caches modules that have been imported.  What happens if you
+type `import zipf` again?
+
+### Exercise: What is `__name__` in different contexts?
+
+1. Start an interactive Python interpreter, and run `print(__name__)`.
+What do you get?
+2. Add the statement `print(__name__)` to the end of the file `zipf.py`
+and from the Python interpreter, run `import zipf`.  What is the name
+printed?
+
+### Exercise: Turn `use.py` into a unit test
```

Comment:
The next chapter will be unit testing, not before. So no reference to unit testing should be included.